File: CommonLibrary/Extensions/ExceptionExtensions.vb
'Gadec Engineerings Software (c) 2022
'Common Library
Imports System.Runtime.CompilerServices

Public Module ExceptionExtensions

    ''' <summary>
    ''' Rethrows the exception.
    ''' </summary>
    ''' <param name="eException"></param>
    <Extension()>
    Public Sub Rethrow(ByVal eException As Exception)
        Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(eException).Throw()
    End Sub

    ''' <summary>
    ''' Adds user-defined information about the exception.
    ''' </summary>
    ''' <param name="eException"></param>
    ''' <param name="text">Information to add.</param>
    <Extension()>
    Public Sub AddData(ByVal eException As Exception, text As String)
        Dim key = "1: "
        Do While eException.Data.Contains(key)
            key = key.AutoNumber
        Loop
        eException.Data.Add(key, text)
    End Sub

End Module

File: GadecLibrary/Extensions/ExceptionExtensions.cs
namespace GadecLibrary.Extensions;
public static class ExceptionExtensions
{
    public static void Rethrow(this Exception eException)
    {
        System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(eException).Throw();
    }

    public static void AddData(this Exception eException, string text)
    {
        var key = "1: ";
        while (eException.Data.Contains(key))
            key = key.AutoNumber;
        eException.Data.Add(key, text);
    }
}
